ext/mysql2/result.c

```diff
@@ -821,12 +821,17 @@ static VALUE rb_mysql_result_each(int argc, VALUE * argv, VALUE self) {
 
   defaults = rb_ivar_get(self, intern_query_options);
   Check_Type(defaults, T_HASH);
-  if (rb_scan_args(argc, argv, "01&", &opts, &block) == 1) {
+  if (rb_scan_args(argc, argv, "01", &opts) == 1) {
     opts = rb_funcall(defaults, intern_merge, 1, opts);
   } else {
     opts = defaults;
   }
 
+  block = Qfalse;
+  if (rb_block_given_p()) {
+    block = Qtrue;
+  }
```

Comment:
Great explanation that actually capturing the block will generate a Proc object that is never called!

To really clarify this code, I suggest a deeper change:

1. Add a comment above `rb_scan_args` that says something like "A block can be passed to this method, but since we don't call the block directly from C, we don't need to capture it into a variable here with the "&" scan arg"

2. Remove the `block` variable entirely.

3. Replace `args.block_given = block;` on line 890 with `args.block_given = rb_block_given_p()`

(Adjust the data type of `args.block_given` to match the return value of `rb_block_given_p()` if needed)

Reply:
@sodabrew Thank you for your review!! I fixed a patch with your idea at https://github.com/brianmario/mysql2/pull/1046/commits/f7b0db4574b2c3b8a92aa298da52a8209c013d8d